infra/cifuzz/filestore/gitlab/__init__.py

```diff
@@ -0,0 +1,121 @@
+import logging
+
+import shutil
+import os
+import json
+import tempfile
+
+import environment
+import http_utils
+import filestore
+
+# pylint:disable=no-self-use,unused-argument
+
+
+class GitlabFilestore(filestore.BaseFilestore):
+  """Implementation of BaseFilestore using Gitlab job artifacts. Relies on
+  having a PRIVATE-TOKEN supplied to access the Gitlab API as seems
+  the only way to download an artifact from another job"""
+
+  def __init__(self, config):
+    super().__init__(config)
+    self.downloaded = set()
+    self.artifacts_dir = environment.get('CFL_ARTIFACTS_DIR', "artifacts")
+    self.download_dir = environment.get('CFL_DOWNLOAD_DIR', "download")
+
+  def _copy_fromdir(self, src, name, reason):
+    destdir = os.path.join(self.config.workspace, self.artifacts_dir, reason,
+                           name)
+    logging.info('Uploading %s to artifacts to %s', reason, destdir)
+    shutil.copytree(src, destdir)
+
+  def upload_crashes(self, name, directory):
+    """Gitlab artifacts implementation of upload_crashes."""
+    self._copy_fromdir(directory, name, "crashes")
+
+  def upload_corpus(self, name, directory, replace=False):
+    """Gitlab artifacts implementation of upload_corpus."""
+    self._copy_fromdir(directory, name, "corpus")
+
+  def upload_build(self, name, directory):
+    """Gitlab artifacts implementation of upload_build."""
+    self._copy_fromdir(directory, name, "build")
+
+  def upload_coverage(self, name, directory):
+    """Gitlab artifacts implementation of upload_coverage."""
+    # would be neat if we could save CI_JOB_ID somewhere,
+    # without needing a token with write-access
+    self._copy_fromdir(directory, name, "coverage")
+
+  def _get_jobid(self, projpath, headers, reason):
+    """Get a specific job id for the latest succesful pipeline
+    with the specific job name."""
+    jobsurl = os.getenv('CI_API_V4_URL') + projpath + "/jobs?scope=success"
+    with tempfile.NamedTemporaryFile() as tmp_file:
+      if not http_utils.download_url(jobsurl, tmp_file.name, headers=headers):
+        logging.error('Failed downloading %s', jobsurl)
+        return None
+      # jq '.[] | select(.name=="clusterfuzzlite-corpus") | .id'
+      with open(tmp_file.name, encoding='ascii') as json_file:
+        data = json.load(json_file)
+        for j in data:
+          # job names are fixed by this
+          if j["name"] == "clusterfuzzlite-" + reason:
+            logging.info('Latest job with %s is %d', reason, j["id"])
+            return j["id"]
+    return None
+
+  def _copy_todir(self, dst, name, reason):
+    if reason not in self.downloaded:
+      branch = os.getenv('CFL_BRANCH')
+      if not branch:
+        branch = os.getenv('CI_DEFAULT_BRANCH')
+      projpath = "/projects/" + os.getenv(
+          'CI_PROJECT_NAMESPACE') + "%2F" + os.getenv('CI_PROJECT_NAME')
+      # headers = {"JOB-TOKEN" : os.getenv('CI_JOB_TOKEN')}
```

Comment:
See https://gitlab.com/gitlab-org/gitlab/-/issues/349316

We can manage with only `JOB-TOKEN`, but to do so, we need to download all artifacts from previous jobs until we find one which matches our expectation (that is one with corpus or coverage...)

Reply:
Oooh this is really bad.
Might be worth skipping this and not implementing the artifact filestore as iirc it is only really needed for saving builds, not for coverage or corpus.

Reply:
This gitlab artifacts filestore works for all 3 : builds, coverage and corpus
As for the Github actions filestore, you need a token, a bit like for the same reason commented below `class GithubActionsFilestore(filestore.BaseFilestore):`
> GitHub actions API does not support downloading artifacts from other runs

Reply:
The github implementation of filestore also supports all 3.
We use git for coverage instead of artifacts so that the report can viewed in browser using github pages.
We use git for corpus instead of artifacts to make it easier to run multiple batch fuzzing instances at once (it's much easier for multiple instances to update a git repo at once than it is if the corpus is a single zip file).

Needing to download all artifacts seems extrememly bad. I don't even know how well this work in the real world. Am I misunderstanding something?

Reply:
> Needing to download all artifacts seems extrememly bad.

That is not done here.
We download only one artifacts archive